docs/RFCS/online_schema_change.md

```diff
@@ -0,0 +1,166 @@
+- Feature Name: online_schema_change
+- Status: draft
+- Start Date: 2015-10-14
+- RFC PR:
+- Cockroach Issue: [#2036](https://github.com/cockroachdb/cockroach/issues/2036)
+
+# Summary
+
+Implement online schema change: adding or removing an index or column
+to a table without blocking access to the table.
+
+# Motivation
+
+The existing schema change operations such as adding or removing an
+index are performed as a single transaction. This approach was
+convenient for initial implementation but infeasible for a table with
+any significant amount of data. Additionally, the single transaction
+approach necessitates reading the table descriptor on every SQL
+operation which is a significant performance bottleneck.
+
+We will implement online schema changed which breaks down a high-level
+schema change operation such as `CREATE INDEX` into a series of
+discrete steps in such a way that user transactions are never blocked
+from accessing the table and yet never leave table or index data in an
+invalid state.
+
+Online schema change will be built on top of [table descriptor
+leases](https://github.com/cockroachdb/cockroach/docs/RFCS/table_descriptor_lease.md)
+which describes the mechanism for asynchronously distributing
+modifications to table descriptors. This RFC is concerned with the
+actual steps of performing a schema change.
+
+# Detailed design
+
+Table descriptors will be enhanced to contain information about an
+on-going schema change operation:
+
+```proto
+message TableDescriptor {
+  ...
+
+  oneof mutation {
+    optional IndexDescriptor add_index;
+    optional IndexDescriptor drop_index;
+    optional ColumnDescriptor add_column;
+    optional ColumnDescriptor drop_column;
+  }
+}
+```
+
+Additionally, index and column descriptors will be enhanced to contain
```

Comment:
Besides what we have below to preserve the state of a mutation, is there a need for the mutation oneof above.